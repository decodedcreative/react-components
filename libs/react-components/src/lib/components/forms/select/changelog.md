# Changelog

## 0.1.0

- [MINOR] Initial component creation. Created styling for disabled and error states
