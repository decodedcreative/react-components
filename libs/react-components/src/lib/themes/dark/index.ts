import colors from './colors';
import components from './components';
import general from './general';
import spacing from './spacing';
import typography from './typography';

const darkTheme = {
  name: 'darkTheme',
  colors,
  components,
  general,
  spacing,
  typography,
};

export default darkTheme;
