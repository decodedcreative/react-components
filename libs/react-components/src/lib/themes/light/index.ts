import colors from './colors';
import components from './components';
import general from './general';
import typography from './typography';

const lightTheme = {
  name: 'lightTheme',
  colors,
  components,
  general,
  typography,
};

export default lightTheme;
