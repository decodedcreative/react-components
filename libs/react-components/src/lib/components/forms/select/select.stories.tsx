import { ComponentStory, ComponentMeta } from '@storybook/react';
import { Select } from './select';

export default {
  title: 'Theme/Forms/Select',
  component: Select,
  parameters: {
    backgrounds: {
      default: 'Light',
      values: [
        { name: 'Light', value: '#ffffff' },
        { name: 'Dark', value: '#161b2f' },
      ],
    },
  },
  argTypes: {},
  args: {
    'aria-invalid': false,
    disabled: false,
  },
  decorators: [
    (Story) => (
      <div style={{ width: '400px' }}>
        <Story />
      </div>
    ),
  ],
} as ComponentMeta<typeof Select>;

const Template: ComponentStory<typeof Select> = ({ children, ...args }) => (
  <Select {...args}>
    <option key={1} value="1">
      Option 1
    </option>
    <option key={2} value="2">
      Option 2
    </option>
    <option key={3} value="3">
      Option 3
    </option>
    <option key={4} value="4">
      Option 4
    </option>
  </Select>
);
export const Default = Template.bind({});
Default.args = {};

export const Error = Template.bind({});
Error.args = {
  'aria-invalid': true,
};

export const Disabled = Template.bind({});
Disabled.args = {
  disabled: true,
};
