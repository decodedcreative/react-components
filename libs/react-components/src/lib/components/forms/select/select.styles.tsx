import styled, { css, ThemeProvider } from 'styled-components';
import { SelectProps } from './select';
import { ChevronDown } from '@styled-icons/boxicons-solid';

export const Select = styled('div')<SelectProps>`
  position: relative;
  width: 100%;
`;

export const SelectInput = styled('select')`
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  cursor: pointer;
  appearance: none;
  width: 100%;
  box-sizing: border-box;
  padding: 11px 16px;
  color: ${(props) => props.theme.colors.color70};
  font-family: ${(props) => props.theme.typography.fontFamily};
  font-size: ${(props) => props.theme.typography.sizes.default};
  line-height: ${(props) => props.theme.typography.lineHeight};
  border: ${(props) => props.theme.general.borderWidth} solid
    ${(props) => props.theme.components.forms.borderColor};

  &:focus::not(&:focus-visible) {
    outline: 0;
    border-color: unset;
  }

  &:focus-visible {
    color: ${(props) => props.theme.colors.color60};
    border-color: ${(props) => props.theme.colors.interactive};
    outline: ${(props) => props.theme.general.focusWidth} solid transparent;
    outline-offset: ${(props) => props.theme.general.focusWidth};
    box-shadow: ${(props) => props.theme.general.focusBoxShadow};
  }

  &:disabled {
    color: ${(props) => props.theme.colors.color50};
  }

  ${(props) =>
    props['aria-invalid'] &&
    css`
      border-color: ${(props) => props.theme.colors.error};
      color: ${(props) => props.theme.colors.error};
    `}
`;

export const SelectMedia = styled('span')`
  position: absolute;
  right: 16px;
  top: 50%;
  transform: translateY(-50%);
  pointer-events: none;
`;

export const SelectIcon = styled(ChevronDown)`
  width: 24px;
  height: 24px;
  color: ${(props) => props.theme.colors.interactive};

  ${SelectInput}:disabled ~ ${SelectMedia} & {
    color: ${(props) => props.theme.colors.color50};
  }

  ${SelectInput}[aria-invalid="true"] ~ ${SelectMedia} & {
    color: ${(props) => props.theme.colors.error};
  }
`;
