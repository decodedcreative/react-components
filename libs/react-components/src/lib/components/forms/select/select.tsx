import { ComponentPropsWithoutRef, forwardRef, ReactNode } from 'react';
import * as S from './select.styles';

export interface SelectInputProps extends ComponentPropsWithoutRef<'select'> {
  children: ReactNode;
}
export interface SelectProps {}

export const Select = forwardRef<
  HTMLSelectElement,
  SelectProps & SelectInputProps
>(({ 'aria-invalid': ariaInvalid, children, required, ...props }, ref) => {
  const wrapperProps = {};

  const inputProps = {
    ...props,
  };

  return (
    <S.Select {...wrapperProps}>
      <S.SelectInput
        aria-invalid={ariaInvalid}
        aria-required={required}
        ref={ref}
        {...inputProps}
      >
        {children}
      </S.SelectInput>
      <S.SelectMedia>
        <S.SelectIcon />
      </S.SelectMedia>
    </S.Select>
  );
});
