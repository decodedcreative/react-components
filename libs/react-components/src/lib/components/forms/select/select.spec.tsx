import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { lightTheme } from '../../../themes';
import { Select, SelectProps } from './select';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

const WithTheme = ({ ...props }) => (
  <ThemeProvider theme={lightTheme}>
    <Select {...props}>
      <option value="1">1</option>
      <option value="2">2</option>
      <option value="3">3</option>
      <option value="4">4</option>
    </Select>
  </ThemeProvider>
);

const createComponentJSX = ({ ...props }: SelectProps) => (
  <WithTheme {...props} />
);
const createComponentRender = ({ ...props }: SelectProps) =>
  render(<WithTheme {...props} />);

describe('Select', () => {
  it('should render successfully', () => {
    const component = createComponentRender({} as SelectProps);
    expect(component).toBeTruthy();
  });

  it('should match the default snapshot', () => {
    const defaultSelect = createComponentJSX({} as SelectProps);
    const defaultSelectJSON = renderer.create(defaultSelect).toJSON();
    expect(defaultSelectJSON).toMatchSnapshot();
  });

  describe('Select - States', () => {
    describe('Select - States - Error', () => {
      it('should have an aria-invalid attribute on its input if in an error state', () => {
        const component = createComponentRender({
          'aria-invalid': true,
          'data-testid': 'select-input',
        } as SelectProps);
        const input = component.getByTestId('select-input');
        expect(input).toHaveAttribute('aria-invalid');
      });
      it('should match the Select With Error snapshot', () => {
        const errorSelect = createComponentJSX({
          'aria-invalid': true,
        } as SelectProps);
        const errorSelectJSON = renderer.create(errorSelect).toJSON();
        expect(errorSelectJSON).toMatchSnapshot();
      });
    });
    describe('Select - States - Disabled', () => {
      it('should disable its input if a disabled prop is supplied', () => {
        const component = createComponentRender({
          disabled: true,
          'data-testid': 'select-input',
        } as SelectProps);
        const input = component.getByTestId('select-input');
        expect(input).toBeDisabled();
      });
      it('should match the Disabled Select snapshot', () => {
        const disabledSelect = createComponentJSX({
          disabled: true,
        } as SelectProps);
        const disabledSelectJSON = renderer.create(disabledSelect).toJSON();
        expect(disabledSelectJSON).toMatchSnapshot();
      });
    });
  });
});
