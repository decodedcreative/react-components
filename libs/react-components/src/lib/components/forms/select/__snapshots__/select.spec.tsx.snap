// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Select - States Select - States - Disabled should match the Disabled Select snapshot 1`] = `
.c5 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  position: relative;
  width: 100%;
}

.c2 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  cursor: pointer;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  width: 100%;
  box-sizing: border-box;
  padding: 11px 16px;
  color: #272727;
  font-family: 'Source Sans Pro',sans-serif;
  font-size: 1rem;
  line-height: 1.5;
  border: 2px solid #8e8e8e;
}

.c2:focus::not(.c1:focus-visible) {
  outline: 0;
  border-color: unset;
}

.c2:focus-visible {
  color: #515151;
  border-color: #006de3;
  outline: 2px solid transparent;
  outline-offset: 2px;
  box-shadow: 0 0 0 2px #ffffff,0 0 0 4px #006de3;
}

.c2:disabled {
  color: #8e8e8e;
}

.c4 {
  position: absolute;
  right: 16px;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  pointer-events: none;
}

.c7 {
  width: 24px;
  height: 24px;
  color: #006de3;
}

.c1:disabled ~ .c3 .c6 {
  color: #8e8e8e;
}

.c1[aria-invalid="true"] ~ .c3 .c6 {
  color: #e90014;
}

<div
  className="c0"
>
  <select
    className="c1 c2"
    disabled={true}
  >
    <option
      value="1"
    >
      1
    </option>
    <option
      value="2"
    >
      2
    </option>
    <option
      value="3"
    >
      3
    </option>
    <option
      value="4"
    >
      4
    </option>
  </select>
  <span
    className="c3 c4"
  >
    <svg
      aria-hidden="true"
      className="c5 c6 c7"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M16.939 7.939 12 12.879l-4.939-4.94-2.122 2.122L12 17.121l7.061-7.06z"
      />
    </svg>
  </span>
</div>
`;

exports[`Select Select - States Select - States - Error should match the Select With Error snapshot 1`] = `
.c5 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  position: relative;
  width: 100%;
}

.c2 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  cursor: pointer;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  width: 100%;
  box-sizing: border-box;
  padding: 11px 16px;
  color: #272727;
  font-family: 'Source Sans Pro',sans-serif;
  font-size: 1rem;
  line-height: 1.5;
  border: 2px solid #8e8e8e;
  border-color: #e90014;
  color: #e90014;
}

.c2:focus::not(.c1:focus-visible) {
  outline: 0;
  border-color: unset;
}

.c2:focus-visible {
  color: #515151;
  border-color: #006de3;
  outline: 2px solid transparent;
  outline-offset: 2px;
  box-shadow: 0 0 0 2px #ffffff,0 0 0 4px #006de3;
}

.c2:disabled {
  color: #8e8e8e;
}

.c4 {
  position: absolute;
  right: 16px;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  pointer-events: none;
}

.c7 {
  width: 24px;
  height: 24px;
  color: #006de3;
}

.c1:disabled ~ .c3 .c6 {
  color: #8e8e8e;
}

.c1[aria-invalid="true"] ~ .c3 .c6 {
  color: #e90014;
}

<div
  className="c0"
>
  <select
    aria-invalid={true}
    className="c1 c2"
  >
    <option
      value="1"
    >
      1
    </option>
    <option
      value="2"
    >
      2
    </option>
    <option
      value="3"
    >
      3
    </option>
    <option
      value="4"
    >
      4
    </option>
  </select>
  <span
    className="c3 c4"
  >
    <svg
      aria-hidden="true"
      className="c5 c6 c7"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M16.939 7.939 12 12.879l-4.939-4.94-2.122 2.122L12 17.121l7.061-7.06z"
      />
    </svg>
  </span>
</div>
`;

exports[`Select should match the default snapshot 1`] = `
.c5 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  position: relative;
  width: 100%;
}

.c2 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  cursor: pointer;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  width: 100%;
  box-sizing: border-box;
  padding: 11px 16px;
  color: #272727;
  font-family: 'Source Sans Pro',sans-serif;
  font-size: 1rem;
  line-height: 1.5;
  border: 2px solid #8e8e8e;
}

.c2:focus::not(.c1:focus-visible) {
  outline: 0;
  border-color: unset;
}

.c2:focus-visible {
  color: #515151;
  border-color: #006de3;
  outline: 2px solid transparent;
  outline-offset: 2px;
  box-shadow: 0 0 0 2px #ffffff,0 0 0 4px #006de3;
}

.c2:disabled {
  color: #8e8e8e;
}

.c4 {
  position: absolute;
  right: 16px;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  pointer-events: none;
}

.c7 {
  width: 24px;
  height: 24px;
  color: #006de3;
}

.c1:disabled ~ .c3 .c6 {
  color: #8e8e8e;
}

.c1[aria-invalid="true"] ~ .c3 .c6 {
  color: #e90014;
}

<div
  className="c0"
>
  <select
    className="c1 c2"
  >
    <option
      value="1"
    >
      1
    </option>
    <option
      value="2"
    >
      2
    </option>
    <option
      value="3"
    >
      3
    </option>
    <option
      value="4"
    >
      4
    </option>
  </select>
  <span
    className="c3 c4"
  >
    <svg
      aria-hidden="true"
      className="c5 c6 c7"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M16.939 7.939 12 12.879l-4.939-4.94-2.122 2.122L12 17.121l7.061-7.06z"
      />
    </svg>
  </span>
</div>
`;
